src/librustc/hir/lowering/expr.rs

```diff
@@ -450,6 +439,16 @@ impl LoweringContext<'_> {
         }
     }
 
+    /// HACK(or_patterns; Centril | dlrobertson): For now we don't push down top level or-patterns
```

Comment:
@dlrobertson We should discuss how to divvy up the remaining work. Most of the changes that you need to do if you push down this change to HIR are not too hard but https://github.com/rust-lang/rust/blob/b50520835f3b14409797139eca23f401a529e711/src/librustc/cfg/construct.rs#L387-L420 seems like an unpleasant one to deal with and probably more relevant to the MIR changes. What I could do (e.g. in a follow up) is to push the hack down to `construct.rs` instead by special casing there + appropriate FIXMEs.

For now I'll go fix the `unused_parens` lint since that is AST based and can be done in parallel with this PR.

Reply:
> We should discuss how to divvy up the remaining work.

Yeah, for now I'm working on getting this working in MIR (so we can start testing).